File: spring/client/src/app/model/review.ts
import { User } from './user';
import { Address } from './address';

export class Review {
  private static default = {
    id: null,
    submitted: new Date().getTime(),
    address: new Address(),
    region: null,
    languages: [],
    sectors: [],
    sectorOther: null,
    cost: null,
    stipend: null,
    workDone: null,
    evaluation: null,
    typicalDay: null,
    difficulties: null,
    safety: null,
    responsiveness: null,
    duration: null,
    other: null,
    reviewerId: null,
    reviewer: null,
    anonymous: true,
    lastEdited: new Date().getTime()
  };
  id: number;
  submitted: number;
  address: Address;
  region: number;
  languages: string[];
  sectors: number[];
  sectorOther: string;
  cost: number;
  stipend: number;
  workDone: string;
  evaluation: string;
  typicalDay: string;
  difficulties: string;
  safety: number;
  responsiveness: number;
  duration: string;
  other: string;
  reviewerId: null;
  reviewer: User;
  anonymous: boolean;
  lastEdited: number;

  constructor(object: any = {}) {
    for (const [key, value] of Object.entries(Review.default)) {
      if (!object[key]) {
        this[key] = value;
      } else if (key === 'address') {
        this[key] = new Address(object[key]);
      } else {
        this[key] = object[key];
      }
    }
  }
}

File: server/src/main/java/edu/berkeley/gpporgs/model/Review.java
package edu.berkeley.gpporgs.model;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.io.Serializable;
import java.util.List;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.Lob;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToOne;
import javax.persistence.Table;

import com.fasterxml.jackson.annotation.JsonIgnore;

/**
 * @author Victor Korir
 * @Date 10/01/2020
 */

@Entity
@Getter
@Setter
@NoArgsConstructor
@Table(name = "reviews")
public class Review implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String created;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    private Organization organization;

    @OneToOne(cascade = CascadeType.ALL)
    private Address address;

    @OneToOne
    private Region region;

    @ManyToMany(cascade = CascadeType.MERGE)
    @JoinTable(
        name = "rev_languages",
        joinColumns = @JoinColumn(name = "rev_id"),
        inverseJoinColumns = @JoinColumn(name = "lan_code")
    )
    private List<Language> languages;

    @Lob
    private String difficulties;

    @ManyToMany(cascade = CascadeType.MERGE)
    @JoinTable(
        name = "rev_sectors",
        joinColumns = @JoinColumn(name = "rev_id"),
        inverseJoinColumns = @JoinColumn(name = "sec_id")
    )
    private List<Sector> sectors;

    private String sectorOther;

    private Integer stipend;

    private Integer cost = 0;

    private String duration;

    @Lob
    private String workDone;

    @Lob
    private String typicalDay;

    @Lob
    private String evaluation;

    @Lob
    private String other;

    private Integer safety = 1;

    private Integer responsiveness = 1;

    private Boolean anonymous = true;

    @ManyToOne
    private User reviewer;
}
